pkg/resources/management.cattle.io/v3/feature/validator_test.go

```diff
@@ -0,0 +1,144 @@
+	featureGVK = metav1.GroupVersionKind{Group: "management.cattle.io", Version: "v3", Kind: "Feature"}
+)
+
+func TestFeatureValueValid(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name       string
+		newFeature v3.Feature
+		oldFeature v3.Feature
+		wantError  bool
+		wantAdmit  bool
+	}{
+		{
+			name: "new feature locked with spec value changed",
+			oldFeature: v3.Feature{
+				Spec: v3.FeatureSpec{
+					Value: admission.Ptr(true),
+				},
+			},
+			newFeature: v3.Feature{
+				Spec: v3.FeatureSpec{
+					Value: admission.Ptr(false),
+				},
+				Status: v3.FeatureStatus{
+					LockedValue: admission.Ptr(true),
+				},
+			},
+			wantAdmit: false,
+		},
+		{
+			name: "new feature not locked with spec value changed",
+			oldFeature: v3.Feature{
+				Spec: v3.FeatureSpec{
+					Value: admission.Ptr(true),
+				},
+			},
+			newFeature: v3.Feature{
+				Spec: v3.FeatureSpec{
+					Value: admission.Ptr(false),
+				},
+				Status: v3.FeatureStatus{
+					LockedValue: admission.Ptr(false),
+				},
+			},
+			wantAdmit: true,
+		},
+		{
+			name: "new feature not locked with spec value unchanged",
+			oldFeature: v3.Feature{
+				Spec: v3.FeatureSpec{
+					Value: admission.Ptr(true),
+				},
+			},
+			newFeature: v3.Feature{
+				Spec: v3.FeatureSpec{
+					Value: admission.Ptr(true),
+				},
+				Status: v3.FeatureStatus{
+					LockedValue: admission.Ptr(true),
+				},
+			},
+			wantAdmit: true,
+		},
+		{
+			name:       "new feature lock is nil",
+			oldFeature: v3.Feature{},
+			newFeature: v3.Feature{},
+			wantAdmit:  true,
+		},
+		{
+			name:       "both feature specs are nil",
+			oldFeature: v3.Feature{},
+			newFeature: v3.Feature{
+				Status: v3.FeatureStatus{
+					LockedValue: admission.Ptr(true),
+				},
+			},
+			wantAdmit: true,
+		},
+		{
+			name:      "bad request",
+			wantError: true,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			admitters := NewValidator().Admitters()
+			assert.Len(t, admitters, 1)
+
+			req := admission.Request{
+				AdmissionRequest: admissionv1.AdmissionRequest{
+					UID:             "2",
+					Kind:            featureGVK,
+					Resource:        featureGVR,
+					RequestKind:     &featureGVK,
+					RequestResource: &featureGVR,
+					Name:            "my-feature",
+					Operation:       admissionv1.Update,
+					UserInfo:        authenicationv1.UserInfo{Username: "test-user", UID: ""},
+					Object:          runtime.RawExtension{},
+					OldObject:       runtime.RawExtension{},
+				},
+			}
+			var err error
+			req.Object.Raw, err = json.Marshal(test.newFeature)
+			assert.NoError(t, err, "Failed to marshal new Feature while creating request")
+			req.OldObject.Raw, err = json.Marshal(test.oldFeature)
+			assert.NoError(t, err, "Failed to marshal old Feature while creating request")
+
+			if test.name == "bad request" {
```

Comment:
I'm not particularly fond of having this custom logic for one of the tests. Maybe add the `AdmissionRequest` as one of the test fields?

Reply:
Agreed, but the setup will be less clear in how the inputs are defined. I am testing features, so my inputs are v3.Feature values. Otherwise, I'd have to define `admissionv1.AdmissionRequest` for each test instead, which would be fine, since the method under test, Admit, actually takes a value of that type. But, again, readability might suffer a bit, since I really care about Features.
It's not ideal, but the test case itself is explicitly declared with the name and expected result.